### DISCUSSION TITLE
Chat within existing notes

### DISCUSSION THREAD

**Discussion by lukemt** [bramses/chatgpt-md#44] [Ideas]
I see many people use chatgpt-md within existing notes and I am wondering if this is an intended use case and what’s actually going on if you do that. Doesn’t chatgpt-md send the entire content of the existing note to ChatGPT as a message and if so are people aware that this happens?

I believe most users that do this would prefer a chatgpt-md section, that is „independent“ of the rest of the note.

What do you guys think about this?

**Comment by sunt05** [comment]
Thank you for initiating this discussion, @lukemt! 

There are two modes to consider:

- Prompt mode: In this mode, a chat divider is inserted to start a prompt chat, such as revising sentences. It does not require much context, as the prompt can be deleted immediately after receiving a satisfactory answer that can be directly merged into the existing note.

- Topical mode: If the prompt chat is insufficient for addressing more complex needs, one may wish to expand a portion of a note into a separate, more comprehensive topical chat. This is similar to the current "chat from highlighted text" feature. However, the existing version loses connection to the original snippets since the new chat is titled with a timestamp rather than some anchor text from the original note. I suggest enhancing this feature by using the first line of the highlighted text as the title for the new chat note. This idea is similar to a feature found in the "note refactor" plugin.

I look forward to exploring more possibilities through this fantastic plugin!

**Reply by lukemt** [reply]
Hi @sunt05 thanks for your participation!

Regarding your first point: Do i understand correctly that you do not want the entire content of the note to be sent to chatgpt (only the prompt beneath the divider)? Or would you prefer the entire note content to be sent to chatgpt (as part of the prompt)?

Regarding the second one: I personally never really made use of the highlight feature but what you are saying sounds reasonable to me. A (back)link from the new chat note to the original note could be useful too.

**Reply by sunt05** [reply]
> Hi @sunt05 thanks for your participation!

Excited to be part of this project!


> 
> Regarding your first point: Do i understand correctly that you do not want the entire content of the note to be sent to chatgpt (only the prompt beneath the divider)? Or would you prefer the entire note content to be sent to chatgpt (as part of the prompt)?
> 

Correct - I'd like only the part beneath the divider to be included  in the prompt. My typical scenario is this: when doing daily journaling, I'd like to enhance part of the daily activities to be a paragraph or just simply add a splash of GPT-ed comments - hence no need to involve the whole daily note. 


> Regarding the second one: I personally never really made use of the highlight feature but what you are saying sounds reasonable to me. A (back)link from the new chat note to the original note could be useful too.
> 

Thanks - looking forward to the enhancement.

**Comment by bramses** [comment]
Hi  [@lukemt](https://github.com/lukemt) and  [@sunt05](https://github.com/sunt05),

Thank you both for the great ideas! My thoughts below:

> I see many people use chatgpt-md within existing notes and I am wondering if this is an intended use case and what’s actually going on if you do that. Doesn’t chatgpt-md send the entire content of the existing note to ChatGPT as a message and if so are people aware that this happens?

I agree, I've also seen many people use ChatGPT MD to enhance their existing notes while having to add minimal changes to their workflow. The way I imagine it, the `chats` folder is meant mainly to add organization to the vault so files aren't floating about everywhere. The `chats` folder also serves as history of chats and a consistent place for long form back and forth conversation with ChatGPT.

> I believe most users that do this would prefer a chatgpt-md section, that is „independent“ of the rest of the note.

This is a good idea, but again, it adds boilerplate people have to think about before "chatting" ergo friction. I prefer opt in over opt out for friction elements (frontmatter required elements for example), and opt out over opt in for convenience elements (auto infer title feature for example)

> Prompt mode: In this mode, a chat divider is inserted to start a prompt chat, such as revising sentences. It does not require much context, as the prompt can be deleted immediately after receiving a satisfactory answer that can be directly merged into the existing note.

You can use the `Add Divider` command to add a new user block in any note.

> Topical mode: If the prompt chat is insufficient for addressing more complex needs, one may wish to expand a portion of a note into a separate, more comprehensive topical chat. This is similar to the current "chat from highlighted text" feature. However, the existing version loses connection to the original snippets since the new chat is titled with a timestamp rather than some anchor text from the original note. I suggest enhancing this feature by using the first line of the highlighted text as the title for the new chat note. This idea is similar to a feature found in the "note refactor" plugin.

This is a good idea, but it breaks the logic `Auto Infer Title` uses (matching automated title regex, or it automatically fails as to not rewrite user written titles)

> Regarding the second one: I personally never really made use of the highlight feature but what you are saying sounds reasonable to me. A (back)link from the new chat note to the original note could be useful too.

I'm considering adding an optional comment block that will be ignored by ChatGPT MD entirely for an upcoming release. Something like `//` or `/* */` in JS, but I have to be careful as to not mess up any existing MD formatting or use characters ChatGPT might insert on it's own accord for code blocks or whatever.

> Correct - I'd like only the part beneath the divider to be included in the prompt. My typical scenario is this: when doing daily journaling, I'd like to enhance part of the daily activities to be a paragraph or just simply add a splash of GPT-ed comments - hence no need to involve the whole daily note.

Hmm, I get how this could be useful, but again it would cause confusion/make the regex extraction really complex. Even if the plugin would say for example see if there is only one divider element and throw out the rest of the file, it would be a case that could break existing chats, or would have to be toggled on and off in settings a ton.

In sum, the main goal is this: "opt in over opt out for friction elements (frontmatter required elements for example), and opt out over opt in for convenience elements (auto infer title feature for example)" so anything that abides by that I'm much more likely to approve!

**Reply by szw** [reply]
>  I'd like only the part beneath the divider to be included in the prompt.

Perhaps it's worth mentioning that ATM, the "prompt mode" sends the whole note to the Chat GPT API anyway. Your note becomes part of the context that way, regardless of dividers. Actually, dividers just split it into more messages. But anything above and beneath them will be sent to Chat GPT.

**Comment by sunt05** [comment]
Thanks @bramses for the prompt follow-up!

Generally I love your philosophy for the design and will practice more in my workflow. 

> > Prompt mode: In this mode, a chat divider is inserted to start a prompt chat, such as revising sentences. It does not require much context, as the prompt can be deleted immediately after receiving a satisfactory answer that can be directly merged into the existing note.
> 
> 
> 
> You can use the `Add Divider` command to add a new user block in any note.
> 

Just want to get some clarification from you - it seems in such scenarios the model is hardcoded to GPT-3.5 and the default front-matter takes no effect - is it so? If so, that's the reason for my feature request #41.

**Comment by szw** [comment]
I added a PR (#47) that allows the "prompt mode" inside a regular note. It will not include anything above the divider - only text beneath it. If no divider is found in a regular note, the call will not happen. If you are in a dedicated chat file, the calls always run, and the entire file serves as context.

**Comment by lukemt** [comment]
Hi @everyone!

Thank you so much for these amazing and detailed responses!

I have an idea, how about this:

Let's imagine we had 3 types of dividers: 
- start 
	-  `<hr class="__chatgpt_plugin_start">`
- separator 
	- `<hr class="__chatgpt_plugin">`
- end
	- `<hr class="__chatgpt_plugin_end">`

If the `start` divider is present, everything above will be ignored by chatgpt-md
Likewise if the `end` divider is present everything beneath will be ignored.
Both the  `start` and `end` dividers would be optional.

Now, if you call the chat command from within a note, it would check whether any of those dividers are present.
If yes, it would continue to chat (above the `end` divider if present, otherwise EOF/cursor).
If not it would insert the following __without__ making an api call:
```
<hr class="__chatgpt_plugin_start">
role::user

[cursor]

<hr class="__chatgpt_plugin_end">
```
Now you would enter your prompt and when you call `chat` again it starts chatting.

---

Benefits: 
- As little friction as possible (the chat command is probably a hotkey)
- Backwards compatibility would be preserved
- Possibility to put headers and footers into your chat templates
- No existing note content would be sent to openai by default

**Reply by lukemt** [reply]
Kinda like this (I drew this with the help of gpt-4 😉)

```mermaid
sequenceDiagram
    participant User
    participant Tool

    Note over User,Tool: Existing note

    User->>Tool: Call chat command (1st time)
    activate Tool
    Tool->>User: Insert section for prompt input
    deactivate Tool

    Note over User,Tool: Prompt input section added to the note
        
    loop Until user stops chatting
    
        User->>User: Enter a prompt in the designated section
        
        User->>Tool: Call chat command (2nd time)
        activate Tool
        
        Note over Tool: Start processing the entered prompt
        
        Tool-->>User: Insert chat response based on given prompt    
	deactivate Tool
	
   end
   
```

**Reply by szw** [reply]
That way, the "call-chat" command won't be idempotent anymore. The same command would either insert the divider for prompt or call the Chat GPT API. And potentially, in longer notes, it could not be immediately apparent if the API would be called or not. And therefore, it might be a less secure approach, IMO. But it's always some tradeoff.

**Reply by lukemt** [reply]
> And therefore, it might be a less secure approach, IMO. 

It would only send the content of the inserted section though. Meaning if the user calls the chat command twice it still does not send any note content.

**Reply by lukemt** [reply]
Actually, at least in theory you won’t really need 3 types of dividers. It would also work with just the one that's already there

**Comment by bramses** [comment]
hey all, please check out the comment block feature i just added! still working towards a solution here about prompt mode vs topical mode, but hopefully this is a step in the right direction: https://github.com/bramses/chatgpt-md/issues/50

**Comment by lukemt** [comment]
Interestingly there is another way to exclude at least header content from sending to OpenAI. Consider this template:

```md
---
...frontmatter...
---

This text will **not** be included in the chat conversation

role::user

Hi

```

If you call the `chat` command here only "Hi" will be sent because `chatgpt-md` will strip away everything before `role::` which is actually pretty convenient :)
